### PR TITLE
Implement SanitizerCoverage support (Refs. #6513)

### PR DESCRIPTION
**Pull request by LebedevRI**
Please refer to https://clang.llvm.org/docs/SanitizerCoverage.html

TLDR: `ModuleSanitizerCoveragePass` instruments the IR by inserting
calls to callbacks at certain constructs. What the callbacks should do
is up to the implementation. They are effectively required for fuzzing
to be effective, and are provided by e.g. libfuzzer.

One huge caveat is `SanitizerCoverageOptions` which controls
which which callbacks should actually be inserted.
I just don't know what to do about it. Right now i have hardcoded
the set that would have been enabled by `-fsanitize=fuzzer-no-link`,
because the alternative, due to halide unflexibility,
would be to introduce ~16 suboptions to control each one.

### CODE REVIEW

**Comment by LebedevRI** [comment]
Hmm, looks like the test needs to be less fragile. Will look into that.

**Comment by alexreinking** [comment]
Just so you know, there's no need to force push... we always squash-merge our PRs and the history can be useful for review.

**Comment by alexreinking** [comment]
Requesting reviews from @steven-johnson since he's familiar with the sanitizer stuff and from @zvookin since he will have opinions on how to deal with passing options to this.

**Comment by LebedevRI** [comment]
> Requesting reviews from @steven-johnson since he's familiar with the sanitizer stuff and from @zvookin since he will have opinions on how to deal with passing options to this.

Thanks!
Re options: we basically only have two choices:
* just add a single best-guess meta-feature, like done here, that hardcodes the set of coverage to enable,
* or add ~16 features to toggle each option in `SanitizerCoverageOptions` struct.

File: test/generator/sancov_aottest.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+extern "C" void __sanitizer_cov_trace_pc_indir(uintptr_t Callee) {
+    if (!enable_callbacks) return;
+    printf("Hit __sanitizer_cov_trace_pc_indir. Success!\n");
+}
+
+template<typename T>
+void clear_out(T &image) {
+    image.for_each_element([&](int x, int y, int c) {
```
Comment (abadams) [inline test/generator/sancov_aottest.cpp:78]: image.fill(-42);

File: src/Module.cpp [hunk 2]
```diff
@@ -846,6 +846,7 @@ void compile_multitarget(const std::string &fn_name,
             Target::MSAN,
             Target::NoRuntime,
             Target::TSAN,
+            Target::SANCOV,
```
Comment (steven-johnson) [inline src/Module.cpp:849]: annoying style nit: all-caps `SANCOV` seems out of place here as an all-caps abbreviation, despite MSAN, etc, since the latter are commonly referred to that way in LLVM docs but the former seems to be `SanitizerCoverage`. If we are using this only for testing purposes then I think we should use either `SanitizerCoverage` or `sanitizer_coverage` to fit in with existing patterns. (Tagging @abadams here for his opinion as well.)

File: test/generator/sancov_aottest.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+extern "C" void __sanitizer_cov_trace_pc_indir(uintptr_t Callee) {
+    if (!enable_callbacks) return;
+    printf("Hit __sanitizer_cov_trace_pc_indir. Success!\n");
+}
+
+template<typename T>
+void clear_out(T &image) {
+    image.fill(-42);
```
Comment (steven-johnson) [inline test/generator/sancov_aottest.cpp:78]: Nit: if `T` is an unsigned type, some compilers can complain about assigning a signed literal to it. Either avoid using a negative value or (probably more simply) just always use a signed type here.

**Comment by steven-johnson** [comment]
> One huge caveat is `SanitizerCoverageOptions` which controls which which callbacks should actually be inserted. I just don't know what to do about it. Right now i have hardcoded the set that would have been enabled by `-fsanitize=fuzzer-no-link`, because the alternative, due to halide unflexibility, would be to introduce ~16 suboptions to control each one.

I think that's probably fine for the time being -- we can expand and add flexibility in the future as desired. That said, we should add a tracking issue here ("Target::SanitizerCoverage only supports equivalent of fuzzer-no-link") and add a link to it in comments at the relevant sites in the code.

**Review by steven-johnson — CHANGES_REQUESTED** [review]
Some nits re: style and comments, otherwise LGTM pending green

File: src/Module.cpp (hunk 2 above)
Comment (abadams) [reply src/Module.cpp:849]: IMO SanitizerCoverage to match the llvm docs, but I don't feel strongly about it

Comment (LebedevRI) [reply src/Module.cpp:849]: Do you want me to change all instances of SANCOV to SanitizerCoverage, or just some?

> If we are using this only for testing purposes

I'm not sure what you mean by that? Roughly, the feature will be requested the way it is requested in the test.

Comment (steven-johnson) [reply src/Module.cpp:849]: > change all instances of SANCOV to SanitizerCoverage, or just some?

I'd vote for all (except the ones in HalideRuntime.h, which should be `sanitizer_coverage` to match the conventions there)

Comment (LebedevRI) [reply src/Module.cpp]: Ok, let me know if i did it wrong.

File: src/Target.cpp [hunk 4]
```diff
@@ -384,7 +384,7 @@ const std::map<std::string, Target::Feature> feature_name_map = {
     {"llvm_large_code_model", Target::LLVMLargeCodeModel},
     {"rvv", Target::RVV},
     {"armv81a", Target::ARMv81a},
-    {"sancov", Target::SANCOV},
+    {"sanitizercoverage", Target::SanitizerCoverage},
```
Comment (steven-johnson) [inline src/Target.cpp:387]: "sanitizer_coverage", so that it matches the name in HalideRuntime.h

File: src/Target.cpp [hunk 5]
```diff
@@ -422,7 +422,7 @@ Target get_jit_target_from_environment() {
     host.set_feature(Target::TSAN);
 #endif
 #if __has_feature(coverage_sanitizer)
-    host.set_feature(Target::SANCOV);
+    host.set_feature(Target::SANITIZERCOVERAGE);
```
Comment (steven-johnson) [inline src/Target.cpp:425]: This looks like it won't compile, did you mean Target::SanitizerCoverage

File: src/Target.cpp [hunk 6]
```diff
@@ -1037,7 +1037,7 @@ bool Target::get_runtime_compatible_target(const Target &other, Target &result)
     }
 
     if ((features & matching_mask) != (other.features & matching_mask)) {
-        Internal::debug(1) << "runtime targets must agree on SoftFloatABI, Debug, TSAN, ASAN, MSAN, HVX, HexagonDma, HVX_shared_object, SANCOV\n"
+        Internal::debug(1) << "runtime targets must agree on SoftFloatABI, Debug, TSAN, ASAN, MSAN, HVX, HexagonDma, HVX_shared_object, SANITIZERCOVERAGE\n"
```
Comment (steven-johnson) [inline src/Target.cpp:1040]: SanitizerCoverage

File: test/generator/sanitizercoverage_generator.cpp [hunk 7]
```diff
@@ -2,12 +2,12 @@
 
 namespace {
 
-class SANCOV : public Halide::Generator<SANCOV> {
+class SANITIZERCOVERAGE : public Halide::Generator<SANITIZERCOVERAGE> {
```
Comment (steven-johnson) [inline test/generator/sanitizercoverage_generator.cpp:5]: s/SANITIZERCOVERAGE/SanitizerCoverage/ or really anything that isn't ALLCAPS

File: test/generator/sanitizercoverage_generator.cpp [hunk 8]
```diff
@@ -2,12 +2,12 @@
...
 public:
     Output<Buffer<int8_t>> output{"output", 3};
 
     void generate() {
-        // Currently the test just exercises Target::SANCOV
+        // Currently the test just exercises Target::SANITIZERCOVERAGE
```
Comment (steven-johnson) [inline test/generator/sanitizercoverage_generator.cpp:10]: Target::SanitizerCoverage

File: test/generator/sanitizercoverage_generator.cpp [hunk 9]
```diff
@@ -16,10 +16,10 @@ class SANCOV : public Halide::Generator<SANCOV> {
     }
 
 private:
-    // Currently the test just exercises Target::SANCOV
+    // Currently the test just exercises Target::SANITIZERCOVERAGE
```
Comment (steven-johnson) [inline test/generator/sanitizercoverage_generator.cpp:19]: Target::SanitizerCoverage

**Review by steven-johnson — CHANGES_REQUESTED** [review]

File: src/Target.cpp (hunk 5 above)
Comment (LebedevRI) [reply src/Target.cpp:425]: I did verify that this builds for me, starting with clean build dir.
